clang/test/CodeGenSYCL/attrs-on-virtual-calls.cpp

```diff
@@ -0,0 +1,45 @@
+// Test verifies that clang codegen properly adds call site attributes to
+// device code
+
+// RUN: %clang_cc1 -triple spir64 -fsycl-allow-virtual-functions \
+// RUN:    -fsycl-is-device -emit-llvm %s -o %t.device
+// RUN: FileCheck %s --input-file=%t.device --check-prefixes=CHECK,CHECK-DEVICE
+// RUN: %clang_cc1 -triple x86_64 -fsycl-allow-virtual-functions \
+// RUN:    -fsycl-is-host -emit-llvm %s -o %t.host
+// RUN: FileCheck %s --input-file=%t.host --check-prefixes=CHECK,CHECK-HOST
+
+// CHECK-LABEL: define {{.*}} @_Z4testv
+// CHECK: call{{.*}}void %[[#]](ptr
+// CHECK-DEVICE-SAME: #[[#ATTRS:]]
+// CHECK-HOST-SAME: %[[#]]){{[[:space:]]}}
+// CHECK: call{{.*}}void @_ZN8Derived17displayEv{{.*}}[[#DIRECT_CALL_ATTRS:]]
+// CHECK-NOT: attributes #[[#DIRECT_CALL_ATTRS]]{{.*}}"virtual-call"
+// CHECK-DEVICE: attributes #[[#ATTRS]] = {{.*}} "virtual-call"
```

Comment:
I'm confused about IR. Is line 12 CHECK here the call which corresponds to virtual call ? If so why is the attribute after the direct call check?

Reply:
> If so why is the attribute after the direct call check?

Line 16 is a `CHECK-NOT`, i.e. we check that attributes list that correspond to a direct call does not contain `"virtual-call"` attribute.